compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -132,122 +136,254 @@ FailureOr<GPUMMASchedule> fitScheduleInSharedMemory(
-    // and then M.
-    if (mTotalTileCount > (warpSqrt * tileSqrt) &&
-        mTotalTileCount % (warpSqrt * tileSqrt) == 0) {
-      mWarpCount = warpSqrt;
-      mTileCount = tileSqrt;
-
-      remainingWarps /= warpSqrt;
-      remainingTiles /= tileSqrt;
-
-      APInt nGCD = GreatestCommonDivisor(APInt(64, nTotalTileCount),
-                                         APInt(64, remainingWarps));
-      nWarpCount = nGCD.getSExtValue();
-      nTotalTileCount /= nWarpCount;
-      remainingWarps /= nWarpCount;
-
-      nGCD = GreatestCommonDivisor(APInt(64, nTotalTileCount),
-                                   APInt(64, remainingTiles));
-      nTileCount = nGCD.getSExtValue();
-    } else {
-      APInt nGCD = GreatestCommonDivisor(APInt(64, nTotalTileCount),
-                                         APInt(64, remainingWarps));
-      nWarpCount = nGCD.getSExtValue();
-      nTotalTileCount /= nWarpCount;
-      remainingWarps /= nWarpCount;
-
-      nGCD = GreatestCommonDivisor(APInt(64, nTotalTileCount),
-                                   APInt(64, remainingTiles));
-      nTileCount = nGCD.getSExtValue();
-      remainingTiles /= nTileCount;
-
-      APInt mGCD = GreatestCommonDivisor(APInt(64, mTotalTileCount),
-                                         APInt(64, remainingWarps));
-      mWarpCount = mGCD.getSExtValue();
-      mTotalTileCount /= mWarpCount;
-      remainingWarps /= mWarpCount;
-
-      mGCD = GreatestCommonDivisor(APInt(64, mTotalTileCount),
-                                   APInt(64, remainingTiles));
-      mTileCount = mGCD.getSExtValue();
+    if (failed(canTargetIntrinsic(pvMatmul, intrinsic, canUpcastAcc,
+                                  mustBeAligned))) {
+      continue;
     }
 
-    const uint64_t kTotalTileCount =
-        llvm::divideCeil(problem.kSize, intrinsic.kSize);
-    APInt kGCD = GreatestCommonDivisor(
-        APInt(64, kTotalTileCount), APInt(64, seeds.bestKTileCountPerSubgroup));
-    int64_t kTileCount = kGCD.getSExtValue();
+    GPUMMASchedule schedule =
+        getOptimalMMASchedule(pvMatmul, intrinsic, pvMatmulSeeds, index);
 
     LLVM_DEBUG({
       llvm::dbgs() << "chosen MMA schedule:\n";
       llvm::dbgs() << "  intrinsic (M, N, K) = (" << intrinsic.mSize << ", "
                    << intrinsic.nSize << ", " << intrinsic.kSize << ")\n";
-      llvm::dbgs() << "  subgroup count (M, N) = (" << mWarpCount << ", "
-                   << nWarpCount << ")\n";
-      llvm::dbgs() << "  subgroup tile count (M, N, K) = (" << mTileCount
-                   << ", " << nTileCount << ", " << kTileCount << ")\n";
+      llvm::dbgs() << "  subgroup count (M, N) = (" << schedule.mWarpCount
+                   << ", " << schedule.nWarpCount << ")\n";
+      llvm::dbgs() << "  subgroup tile count (M, N, K) = ("
+                   << schedule.mTileCount << ", " << schedule.nTileCount << ", "
+                   << schedule.kTileCount << ")\n";
```

Comment:
Also here

Reply:
removed the intrinsic printing, and just using schedule printing now.